### PR TITLE
Add Executor.Priority.BACKGROUND

### PR DESCRIPTION
**Pull request by joshmoore**
The addition of parallel import has made it quite easy
to use up the omero.db.pool_size number of PostgreSQL
connections available to the server. To prevent this,
the ManagedImportRequests that are submitted on import
now go to a lower priority pool of threads.

# Testing this PR

1. with parallel-fileset, import a number of filesets concurrently
2. watch that other clients/users can still login regardless of the import progress
3. if desired, set `omero.threads.background_threads` to one and watch that import is *serial*
4. if desired, set `omero.threads.background_timeout` to something small to see client errors

### CODE REVIEW

File: components/server/src/ome/services/util/Executor.java [hunk 1]
```diff
@@ -354,6 +364,11 @@ public Impl(CurrentDetails principalHolder, SessionFactory factory,
             this.service = service;
             // Allowed to create more threads.
             this.systemService = Executors.newCachedThreadPool();
+            this.backgroundService = Executors.newFixedThreadPool(10,
```
Comment (mtbc) [inline components/server/src/ome/services/util/Executor.java]: Would be great if this were configurable (so we can experiment) and maybe wise to default to less than default DB pool size.

Comment (joshmoore) [reply components/server/src/ome/services/util/Executor.java]: Agreed. This is very much a temporary fix for testing purposes while we figure out the appropriate concurrency abstraction.

**Comment by joshmoore** [comment]
The loggers which occur in the new background threads:

```
[hudson@ome-c6100-1 dist]$ grep ground var/log/Blitz-0.log  | cut -b34-71 | sort | uniq -c
      8             loci.formats.FormatHandler
     11               loci.formats.ImageReader
     44                  loci.formats.Memoizer
     20   ome.formats.enums.IQueryEnumProvider
     22      ome.formats.importer.ImportConfig
     44   ome.formats.OMEROMetadataStoreClient
     11        ome.security.basic.EventHandler
     33       ome.services.util.ServiceHandler
    187                org.perf4j.TimingLogger
     33   o.s.blitz.repo.ManagedImportRequestI
      8 o.s.b.r.ManagedImportRequestI.@448815d  (x 10)
```

**Comment by joshmoore** [comment]
Primary questions for the current implementation in this PR are:

 * Do we need to increase pool size (DB connections) by default?
 * What percentage of max_threads should background_threads be by default?
 * How long should we wait to _submit_ an import before throwing to the user?

cc: @dominikl @sbesson @mtbc @rgozim @jburel

**Comment by joshmoore** [comment]
Note: configuration property changed in description

Timeout is now configurable (in ms) with `omero.threads.background_timeout`. Setting it to 1 I see:

```
/tmp/fakes/image_005.fake	2018-08-31 13:07:44,018 50017      [      main] INFO   ormats.importer.cli.LoggingImportMonitor - FILESET_UPLOAD_END
/tmp/fakes/image_005.fake	2018-08-31 13:07:44,025 50024      [      main] ERROR        ome.formats.importer.ImportLibrary - Error on import: Failed to execute 1 after ome.services.util.Executor$Impl$1@7928546fms
/tmp/fakes/image_005.fake	omero.InternalException: null
/tmp/fakes/image_005.fake		at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method) ~[na:1.8.0_25]
/tmp/fakes/image_005.fake		at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62) ~[na:1.8.0_25]
/tmp/fakes/image_005.fake		at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45) ~[na:1.8.0_25]
/tmp/fakes/image_005.fake		at java.lang.reflect.Constructor.newInstance(Constructor.java:408) ~[na:1.8.0_25]
/tmp/fakes/image_005.fake		at java.lang.Class.newInstance(Class.java:438) ~[na:1.8.0_25]
/tmp/fakes/image_005.fake		at IceInternal.BasicStream.createUserException(BasicStream.java:2779) ~[ice.jar:na]
/tmp/fakes/image_005.fake		at IceInternal.BasicStream.access$300(BasicStream.java:14) ~[ice.jar:na]
/tmp/fakes/image_005.fake		at IceInternal.BasicStream$EncapsDecoder10.throwException(BasicStream.java:3298) ~[ice.jar:na]
/tmp/fakes/image_005.fake		at IceInternal.BasicStream.throwException(BasicStream.java:2291) ~[ice.jar:na]
/tmp/fakes/image_005.fake		at IceInternal.OutgoingAsync.throwUserException(OutgoingAsync.java:399) ~[ice.jar:na]
/tmp/fakes/image_005.fake		at omero.grid.ImportProcessPrxHelper.end_verifyUpload(ImportProcessPrxHelper.java:1616) ~[blitz.jar:na]
/tmp/fakes/image_005.fake		at omero.grid.ImportProcessPrxHelper.verifyUpload(ImportProcessPrxHelper.java:1495) ~[blitz.jar:na]
/tmp/fakes/image_005.fake		at omero.grid.ImportProcessPrxHelper.verifyUpload(ImportProcessPrxHelper.java:1482) ~[blitz.jar:na]
/tmp/fakes/image_005.fake		at ome.formats.importer.ImportLibrary.importImage(ImportLibrary.java:523) ~[blitz.jar:na]
/tmp/fakes/image_005.fake		at ome.formats.importer.ImportLibrary.importCandidates(ImportLibrary.java:289) ~[blitz.jar:na]
/tmp/fakes/image_005.fake		at ome.formats.importer.cli.CommandLineImporter.start(CommandLineImporter.java:289) [blitz.jar:na]
/tmp/fakes/image_005.fake		at ome.formats.importer.cli.CommandLineImporter.main(CommandLineImporter.java:967) [blitz.jar:na]
/tmp/fakes/image_005.fake	2018-08-31 13:07:44,026 50025      [      main] INFO         ome.formats.importer.ImportLibrary - Exiting on error
/tmp/fakes/image_005.fake
/tmp/fakes/image_005.fake	==> Summary
/tmp/fakes/image_005.fake	1 file uploaded, 0 filesets created, 0 images imported, 0 errors in 0:00:26.423
```

File: components/server/src/ome/services/scheduler/ThreadPool.java [hunk 2]
```diff
@@ -7,46 +7,173 @@
...
-                TimeUnit.MILLISECONDS, queue); // factory
     }
 
+    public ThreadPool(int minThreads, int maxThreads, long msTimeout,
+            int backgroundThreads, long backgroundTimeout) {
+        super(minThreads, maxThreads, msTimeout, TimeUnit.MILLISECONDS,
+                new LinkedBlockingQueue<Runnable>());
+        this.maxBackground = new Semaphore(backgroundThreads);
+        this.backgroundTimeout = backgroundTimeout;
+        log.info("ThreadPool: normal={}, background={}, timeout={}",
```
Comment (mtbc) [inline components/server/src/ome/services/scheduler/ThreadPool.java]: may be confused with `msTimeout`

Comment (joshmoore) [reply components/server/src/ome/services/scheduler/ThreadPool.java]: I _knew_ I should have taken the time to clarify that... ;)

**Comment by mtbc** [comment]
Guessing that any accompanying documentation can be on top of https://github.com/openmicroscopy/ome-documentation/pull/1898.

File: components/server/src/ome/services/scheduler/ThreadPool.java [hunk 3]
```diff
@@ -7,46 +7,173 @@
...
-    private final ThreadFactory factory;
+    /**
+     * Milliseconds to wait until calls to {@link #background(Callable)} will
+     * timeout.
+     */
+    private final long backgroundTimeout;
 
-    private final ExecutorService executor;
+    public ThreadPool() {
+        // Values from Executres.newCachedThreadPool
```
Comment (mtbc) [inline components/server/src/ome/services/scheduler/ThreadPool.java]: typo

File: components/server/src/ome/services/scheduler/ThreadPool.java [hunk 4]
```diff
@@ -7,46 +7,173 @@
...
+     * Standard {@link ThreadPoolExecutor} extension point which checks for the
+     * BackgroundFutureTask marker and releases a slot in the {@link #maxBackground}
+     * {@link Semaphore}.
+     */
+    @Override
+    protected void afterExecute(Runnable r, Throwable t) {
+        try {
+            if (r instanceof BackgroundFutureTask) {
+                maxBackground.release();
+                ((BackgroundFutureTask) r).sw.stop("omero.background.task");
```
Comment (mtbc) [inline components/server/src/ome/services/scheduler/ThreadPool.java:134]: I'm trying to figure out where this stopwatch is started. :smiley:

File: etc/omero.properties [hunk 5]
```diff
@@ -57,6 +57,11 @@ omero.db.authority=export.openmicroscopy.org
 # will be used by OMERO. Your database installation
 # will need to be configured to accept *at least* as
 # many, preferably more, connections as this value.
+#
+# The values of omero.threads.max_threads and
+# omero.threads.background_threads play a role
+# in how many DB connections are used at any given
+# time.
```
Comment (mtbc) [inline etc/omero.properties]: Maybe advise to set higher than `omero.threads.background_threads` and default to such?

File: components/server/src/ome/services/scheduler/ThreadPool.java (hunk 4 above)
Comment (joshmoore) [reply components/server/src/ome/services/scheduler/ThreadPool.java:134]: line 172 (https://github.com/openmicroscopy/openmicroscopy/pull/5839/files/5e501abf68bea82dde8895827a99bb44b35bd70f#diff-12c557a63afd55b5d4bb0e80d1287100R172 ) or you mean -- "documentation please" ? :smile:

Comment (mtbc) [reply components/server/src/ome/services/scheduler/ThreadPool.java:134]: Heh, no, it's fine, I just didn't realize or remember than simply creating it starts it. (Maybe need to get more sleep!)

File: components/server/src/ome/services/scheduler/ThreadPool.java [hunk 6]
```diff
@@ -7,46 +7,173 @@
...
 import java.util.concurrent.ThreadPoolExecutor;
 import java.util.concurrent.TimeUnit;
 
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-import org.quartz.Scheduler;
+import org.perf4j.StopWatch;
+import org.perf4j.slf4j.Slf4JStopWatch;
+
+import com.google.common.util.concurrent.ThreadFactoryBuilder;
```
Comment (mtbc) [inline components/server/src/ome/services/scheduler/ThreadPool.java]: Don't think we use this.

File: etc/omero.properties (hunk 5 above)
Comment (joshmoore) [reply etc/omero.properties]: Better, @mtbc?

Comment (mtbc) [reply etc/omero.properties]: Absolutely. :+1:

**Comment by mtbc** [comment]
With `omero.threads.background_threads=2` and `--parallel-{upload,fileset}=3` I do see that this PR changes output like,
```
2018-09-03 13:31:38,395 10331      [l.Client-0] INFO   ormats.importer.cli.LoggingImportMonitor - METADATA_IMPORTED Step: 1 of 5  Logfile: 2739
2018-09-03 13:31:38,502 10438      [l.Client-2] INFO   ormats.importer.cli.LoggingImportMonitor - METADATA_IMPORTED Step: 1 of 5  Logfile: 2740
2018-09-03 13:31:38,685 10621      [l.Client-3] INFO   ormats.importer.cli.LoggingImportMonitor - METADATA_IMPORTED Step: 1 of 5  Logfile: 2743
2018-09-03 13:31:38,786 10722      [l.Client-1] INFO   ormats.importer.cli.LoggingImportMonitor - PIXELDATA_PROCESSED Step: 2 of 5  Logfile: 2739
2018-09-03 13:31:38,902 10838      [l.Client-0] INFO   ormats.importer.cli.LoggingImportMonitor - PIXELDATA_PROCESSED Step: 2 of 5  Logfile: 2740
2018-09-03 13:31:39,019 10955      [l.Client-2] INFO   ormats.importer.cli.LoggingImportMonitor - THUMBNAILS_GENERATED Step: 3 of 5  Logfile: 2739
2018-09-03 13:31:39,027 10963      [l.Client-3] INFO   ormats.importer.cli.LoggingImportMonitor - PIXELDATA_PROCESSED Step: 2 of 5  Logfile: 2743
2018-09-03 13:31:39,053 10989      [l.Client-1] INFO   ormats.importer.cli.LoggingImportMonitor - METADATA_PROCESSED Step: 4 of 5  Logfile: 2739
2018-09-03 13:31:39,078 11014      [l.Client-0] INFO   ormats.importer.cli.LoggingImportMonitor - OBJECTS_RETURNED Step: 5 of 5  Logfile: 2739
2018-09-03 13:31:39,153 11089      [l.Client-2] INFO   ormats.importer.cli.LoggingImportMonitor - THUMBNAILS_GENERATED Step: 3 of 5  Logfile: 2740
```
to being like,
```
2018-09-03 13:42:56,332 10062      [l.Client-1] INFO   ormats.importer.cli.LoggingImportMonitor - PIXELDATA_PROCESSED Step: 2 of 5  Logfile: 2789
2018-09-03 13:42:56,375 10105      [l.Client-0] INFO   ormats.importer.cli.LoggingImportMonitor - PIXELDATA_PROCESSED Step: 2 of 5  Logfile: 2790
2018-09-03 13:42:56,392 10122      [3-thread-2] INFO   ormats.importer.cli.LoggingImportMonitor - FILE_UPLOAD_COMPLETE: /tmp/dicom/brain-8.dcm
2018-09-03 13:42:56,702 10432      [l.Client-2] INFO   ormats.importer.cli.LoggingImportMonitor - THUMBNAILS_GENERATED Step: 3 of 5  Logfile: 2789
2018-09-03 13:42:56,702 10432      [l.Client-1] INFO   ormats.importer.cli.LoggingImportMonitor - THUMBNAILS_GENERATED Step: 3 of 5  Logfile: 2790
2018-09-03 13:42:56,771 10501      [l.Client-0] INFO   ormats.importer.cli.LoggingImportMonitor - METADATA_PROCESSED Step: 4 of 5  Logfile: 2790
2018-09-03 13:42:56,771 10501      [l.Client-2] INFO   ormats.importer.cli.LoggingImportMonitor - METADATA_PROCESSED Step: 4 of 5  Logfile: 2789
2018-09-03 13:42:56,825 10555      [l.Client-1] INFO   ormats.importer.cli.LoggingImportMonitor - OBJECTS_RETURNED Step: 5 of 5  Logfile: 2790
2018-09-03 13:42:56,834 10564      [l.Client-0] INFO   ormats.importer.cli.LoggingImportMonitor - OBJECTS_RETURNED Step: 5 of 5  Logfile: 2789
```
which makes sense.

**Pull request merged**